core/state/statedb.go

```diff
@@ -374,14 +388,95 @@ func (s *StateDB) MergeSlotDB(slotDb *StateDB, slotReceipt *types.Receipt, txInd
 		if _, exist := s.stateObjectsDirty[addr]; !exist {
 			s.stateObjectsDirty[addr] = struct{}{}
 		}
-
+		// system address is EOA account, it should have no storage change
 		if addr == systemAddress {
 			continue
 		}
 
 		// stateObjects: KV, balance, nonce...
-		if obj, ok := slotDb.getStateObjectFromStateObjects(addr); ok {
-			s.storeStateObjectToStateDB(addr, obj.deepCopyForSlot(s))
+		dirtyObj, ok := slotDb.getStateObjectFromStateObjects(addr)
+		if !ok {
+			panic(fmt.Sprintf("MergeSlotDB dirty object not exist! (txIndex: %d, addr: %s)", slotDb.txIndex, addr.String()))
+		}
+		mainObj, exist := s.loadStateObjectFromStateDB(addr)
+
+		log.Debug("MergeSlotDB", "txIndex", slotDb.txIndex, "addr", addr,
+			"exist", exist, "dirtyObj.deleted", dirtyObj.deleted)
+		if !exist {
+			// addr not exist on main DB, do ownership transfer
+			dirtyObj.db = s
+			dirtyObj.finalise(true) // prefetch on dispatcher
+			s.storeStateObjectToStateDB(addr, dirtyObj)
+			delete(slotDb.parallel.dirtiedStateObjectsInSlot, addr) // transfer ownership
+		} else {
+			// addr already in main DB, do merge: balance, KV, code, State(create, suicide)
+			// can not do copy or ownership transfer directly, since dirtyObj could have outdated
+			// data(may be update within the conflict window)
+
+			// This is debug log for add balance with 0, can be removed
+			if dirtyObj.deleted && dirtyObj.empty() && !mainObj.empty() {
+				if _, exist := slotDb.parallel.balanceChangedInSlot[addr]; !exist {
+					// add(0) could trigger empty delete
+					// note: what if add/sub with 0 result, it is also a fake delete?
+					//       it is ok, since none-zero add will produce a read record, should conflict
+					log.Warn("MergeSlotDB empty deleted", "txIndex", slotDb.txIndex, "addr", addr)
+				}
+			}
+
+			// Do deepCopy a temporary *StateObject for safety,
+			// since slot could read the address, dispatch should avoid overwrite the StateObject directly
+			// otherwise, it could crash for: concurrent map iteration and map write
+			var newMainObj *StateObject
+			if _, suicided := slotDb.parallel.stateObjectSuicided[addr]; suicided {
+				if !dirtyObj.suicided {
+					// debug purpose, can be removed
+					log.Warn("MergeSlotDB suicide and recreated", "txIndex", slotDb.txIndex, "addr", addr)
+				}
+			}
+
+			if dirtyObj.deleted {
+				log.Debug("MergeSlotDB state object merge: Suicide")
+				if !dirtyObj.suicided && !dirtyObj.empty() {
+					// none suicide object, should be empty delete
+					log.Error("MergeSlotDB none suicide deleted, should be empty", "txIndex", slotDb.txIndex, "addr", addr)
+				}
+				// suicided object will keep its worldstate(useless), and will be deleted from trie on block commit
+				newMainObj = dirtyObj.deepCopy(s)
+				delete(s.snapAccounts, addr)
```

Comment:
<img width="537" alt="image" src="https://user-images.githubusercontent.com/5030187/156991633-83a01e90-d610-4208-9bc8-c9219b0f838f.png">

there is only one place which sets `dirtyObj.deleted` == true. so we do not need to add 
```
delete(s.snapAccounts, addr)
delete(s.snapStorage, addr)
```
here. see the below comment, we can add these below.

Reply:
for createObject() will add an addr for snapDestructs to destroy previous object, while it need to keep the addr in snapAccounts & snapStorage.

We can not simply do:
		for k := range slotDb.snapDestructs {
			s.snapDestructs[k] = struct{}{}
		        delete(s.snapAccounts, k)
			delete(s.snapStorage, k)
		}